libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt

```diff
@@ -791,21 +799,33 @@ class MapboxNavigation @VisibleForTesting internal constructor(
             routes,
             initialLegIndex,
             reason,
+            callback,
         )
     }
 
     private fun internalSetNavigationRoutes(
         routes: List<NavigationRoute>,
         legIndex: Int = 0,
         @RoutesExtra.RoutesUpdateReason reason: String,
+        callback: RoutesSetCallback? = null,
     ) {
         rerouteController?.interrupt()
         restartRouteScope()
         threadController.getMainScopeAndRootJob().scope.launch(Dispatchers.Main.immediate) {
             routeUpdateMutex.withLock {
                 navigationSession.setRoutes(routes)
-                setRoutesToTripSession(routes, legIndex, reason)
-                directionsSession.setRoutes(routes, legIndex, reason)
+                when (val processedRoutes = setRoutesToTripSession(routes, legIndex, reason)) {
+                    is NativeSetRouteValue -> {
+                        directionsSession.setRoutes(routes, legIndex, reason)
+                        callback?.onRoutesSetResult(RoutesSetCallbackSuccess(routes))
+                    }
+                    is NativeSetRouteError -> {
+                        logW("Routes $routes will be ignored as they are not valid")
+                        callback?.onRoutesSetError(
+                            RoutesSetCallbackError(routes, processedRoutes.error)
+                        )
```

Comment:
I think it would be nice to have an instrumentation test for this case. WDYT?

Reply:
It turned out for now there's no such case when route is parsed successfully but `setRoutes` returns the error. However, it can filter out some of the alternatives if they have less than 3 intersections. Do you think the behaviour now is OK? Practically, SDK will never invoke `onRoutesSetError`. We could check if alternatives were filtered out, if the result list is empty and so on. But probably it's better to check for the error, just like it's implemented now. WDYT? @RingerJK @VysotskiVadim

Reply:
After some discussion with NN we found a case when it could happen: wrong leg index. They also said that the logic of setting routes is still pretty complex and they can not guarantee always good result. So it makes sense to handle errors.

Reply:
> We could check if alternatives were filtered out, if the result list is empty and so on.

I like this idea. It works if NN explicitly provides this info to the platform in the callback, then we don't have concerns. I think we should be good to keep API as + cut a ticket for NN and track it separately

Reply:
Created an issue and cc'ed you.